File: examples/clean_api_python/application/commands/create_product.py
"""CreateProduct command + handler — Application / Commands"""
from dataclasses import dataclass
from typing import Dict, Any
import sys, os, time

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '../../../../..')))

from backbone.application.exceptions import ValidationException
from backbone.infrastructure.logging import LoggerFactory
from examples.clean_api_python.domain.entities.product import Product, ValidationError
from examples.clean_api_python.domain.repositories.product_repository import ProductRepository


@dataclass
class CreateProductCommand:
    name: str
    description: str
    price: float
    category: str
    stock: int


@dataclass
class CreateProductResult:
    product_id: str


class CreateProductCommandHandler:

    def __init__(self, repository: ProductRepository):
        self.repository = repository
        self.logger = LoggerFactory.create_layer_logger(
            "product-api", "application", "CreateProductCommandHandler"
        )

    def handle(self, command: CreateProductCommand, context: Dict[str, Any]) -> CreateProductResult:
        self.logger.info("Handling CreateProductCommand", extra_data={
            "name": command.name, "category": command.category, "price": command.price,
        })

        self._validate(command)

        try:
            product = Product(
                name=command.name,
                description=command.description,
                category=command.category,
                price=command.price,
                stock=command.stock,
            )
        except ValidationError as e:
            raise ValidationException("Product validation failed", [{"field": e.field, "message": e.message}])

        start = time.time()
        self.repository.create(product)
        self.logger.info("Product created", extra_data={
            "product_id": product.id, "duration_ms": int((time.time() - start) * 1000),
        })

        return CreateProductResult(product_id=product.id)

    def _validate(self, command: CreateProductCommand) -> None:
        if not command.name:
            raise ValueError("name is required")
        if not command.category:
            raise ValueError("category is required")
        if command.price <= 0:
            raise ValueError("price must be greater than 0")
        if command.stock < 0:
            raise ValueError("stock cannot be negative")

File: backbone-go/examples/clean-api-go/application/commands/create_product.go
// Package commands contains write-side application handlers (CQRS).
package commands

import (
	"context"
	"fmt"
	"time"

	"github.com/freakjazz/backbone-go/application/exceptions"
	"github.com/freakjazz/backbone-go/examples/clean-api-go/domain/entities"
	"github.com/freakjazz/backbone-go/examples/clean-api-go/domain/repositories"
	"github.com/freakjazz/backbone-go/infrastructure/logging"
)

// CreateProductCommand is the write input for creating a product.
type CreateProductCommand struct {
	Name        string  `json:"name"`
	Description string  `json:"description"`
	Price       float64 `json:"price"`
	Category    string  `json:"category"`
	Stock       int     `json:"stock"`
}

// CreateProductResult is the write output.
type CreateProductResult struct {
	ProductID string
}

// CreateProductCommandHandler handles the CreateProduct command.
type CreateProductCommandHandler struct {
	repository repositories.ProductRepository
	logger     *logging.EnhancedLogger
}

// NewCreateProductCommandHandler creates a new handler instance.
func NewCreateProductCommandHandler(repo repositories.ProductRepository, logger *logging.EnhancedLogger) *CreateProductCommandHandler {
	return &CreateProductCommandHandler{repository: repo, logger: logger}
}

// Handle executes the command.
func (h *CreateProductCommandHandler) Handle(ctx context.Context, cmd CreateProductCommand) (*CreateProductResult, error) {
	log := h.logger.
		WithLayer("application").
		WithHandler("CreateProductCommandHandler").
		WithMethod("Handle").
		WithContext(map[string]interface{}{"request_id": ctx.Value("request_id")})

	if err := h.validate(cmd); err != nil {
		log.ErrorWithCode("Validation failed", 10001001, map[string]interface{}{"error": err.Error()})
		return nil, err
	}

	product, err := entities.NewProduct(cmd.Name, cmd.Description, cmd.Category, cmd.Price, cmd.Stock)
	if err != nil {
		log.ErrorWithCode("Domain validation failed", 10001002, map[string]interface{}{"error": err.Error()})
		return nil, exceptions.NewValidationException("Product validation failed",
			[]exceptions.ValidationError{{Field: "product", Message: err.Error()}})
	}

	start := time.Now()
	if err := h.repository.Create(ctx, product); err != nil {
		log.ErrorWithCode("Failed to persist product", 10001003, map[string]interface{}{
			"error": err.Error(), "duration_ms": time.Since(start).Milliseconds(),
		})
		return nil, exceptions.NewUseCaseException("Failed to create product", "CreateProductCommandHandler")
	}

	log.Info("Product created", map[string]interface{}{
		"product_id": product.ID, "duration_ms": time.Since(start).Milliseconds(),
	})
	return &CreateProductResult{ProductID: product.ID}, nil
}

func (h *CreateProductCommandHandler) validate(cmd CreateProductCommand) error {
	if cmd.Name == "" {
		return fmt.Errorf("name is required")
	}
	if cmd.Category == "" {
		return fmt.Errorf("category is required")
	}
	if cmd.Price <= 0 {
		return fmt.Errorf("price must be greater than 0")
	}
	if cmd.Stock < 0 {
		return fmt.Errorf("stock cannot be negative")
	}
	return nil
}
